File: services/modelRuntime/providerPolicy.ts
import { getCustomBackendProtocol } from '../../utils/customBackend.js'
import type {
  OfficialSemanticCapabilities,
  ThinkingParityStrategy,
} from './canonical.js'

export type ProviderModelPolicy = {
  capabilities: OfficialSemanticCapabilities
  syntheticTags: {
    responseClose: string
    responseOpen: string
    thinkingClose: string
    thinkingOpen: string
  }
  thinkingStrategy: ThinkingParityStrategy
}

const MOSSEN_COMPATIBLE_CAPABILITIES: OfficialSemanticCapabilities = {
  mixedContentToolUse: true,
  nativeThinkingBlocks: true,
  reasoningBudget: true,
  streamingToolArgDeltas: true,
  structuredStopReasons: true,
  supportsAssistantPreludeBeforeToolUse: true,
  toolCallArgsEncoding: 'object',
  toolResultRoleStyle: 'mossen_user_tool_result',
}

const OPENAI_COMPATIBLE_CAPABILITIES: OfficialSemanticCapabilities = {
  mixedContentToolUse: false,
  nativeThinkingBlocks: false,
  reasoningBudget: false,
  streamingToolArgDeltas: true,
  structuredStopReasons: false,
  supportsAssistantPreludeBeforeToolUse: true,
  toolCallArgsEncoding: 'json_string',
  toolResultRoleStyle: 'openai_tool_role',
}

const DEFAULT_SYNTHETIC_TAGS = {
  thinkingOpen: '<assistant_thinking>',
  thinkingClose: '</assistant_thinking>',
  responseOpen: '<assistant_response>',
  responseClose: '</assistant_response>',
} as const

function resolveThinkingStrategy(
  requestedThinking: unknown,
  protocol: string,
): ThinkingParityStrategy {
  if (protocol === 'mossen-compatible') {
    return 'native'
  }

  if (
    requestedThinking &&
    typeof requestedThinking === 'object' &&
    'type' in requestedThinking &&
    (requestedThinking as { type?: string }).type === 'disabled'
  ) {
    return 'none'
  }

  const envStrategy =
    process.env.MOSSEN_CODE_CUSTOM_THINKING_PARITY_STRATEGY?.trim()
  if (envStrategy === 'synthetic_two_pass') {
    return 'synthetic_two_pass'
  }
  if (envStrategy === 'none') {
    return 'none'
  }

  return 'synthetic_single_pass'
}

export function getOfficialSemanticCapabilities(): OfficialSemanticCapabilities {
  return getCustomBackendProtocol() === 'openai-compatible'
    ? OPENAI_COMPATIBLE_CAPABILITIES
    : MOSSEN_COMPATIBLE_CAPABILITIES
}

export function resolveProviderModelPolicy({
  requestedThinking,
}: {
  requestedThinking: unknown
}): ProviderModelPolicy {
  const protocol = getCustomBackendProtocol()
  return {
    capabilities:
      protocol === 'openai-compatible'
        ? OPENAI_COMPATIBLE_CAPABILITIES
        : MOSSEN_COMPATIBLE_CAPABILITIES,
    syntheticTags: { ...DEFAULT_SYNTHETIC_TAGS },
    thinkingStrategy: resolveThinkingStrategy(requestedThinking, protocol),
  }
}

File: crates/mossen-agent/src/services/model_runtime/provider_policy.rs
//! Provider policy — determines semantic capabilities based on backend protocol.

use serde::{Deserialize, Serialize};

use super::canonical::{
    OfficialSemanticCapabilities, ThinkingParityStrategy, ToolCallArgsEncoding, ToolResultRoleStyle,
};

/// Provider model policy with capabilities and thinking strategy.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProviderModelPolicy {
    pub capabilities: OfficialSemanticCapabilities,
    pub synthetic_tags: SyntheticTags,
    pub thinking_strategy: ThinkingParityStrategy,
}

/// Synthetic XML tags used for non-native thinking.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SyntheticTags {
    pub response_close: String,
    pub response_open: String,
    pub thinking_close: String,
    pub thinking_open: String,
}

impl Default for SyntheticTags {
    fn default() -> Self {
        Self {
            thinking_open: "<assistant_thinking>".to_string(),
            thinking_close: "</assistant_thinking>".to_string(),
            response_open: "<assistant_response>".to_string(),
            response_close: "</assistant_response>".to_string(),
        }
    }
}

/// Mossen-native (Mossen API) capabilities.
fn mossen_compatible_capabilities() -> OfficialSemanticCapabilities {
    OfficialSemanticCapabilities {
        mixed_content_tool_use: true,
        native_thinking_blocks: true,
        reasoning_budget: true,
        streaming_tool_arg_deltas: true,
        structured_stop_reasons: true,
        supports_assistant_prelude_before_tool_use: true,
        tool_call_args_encoding: ToolCallArgsEncoding::Object,
        tool_result_role_style: ToolResultRoleStyle::MossenUserToolResult,
    }
}

/// OpenAI-compatible capabilities.
fn openai_compatible_capabilities() -> OfficialSemanticCapabilities {
    OfficialSemanticCapabilities {
        mixed_content_tool_use: false,
        native_thinking_blocks: false,
        reasoning_budget: false,
        streaming_tool_arg_deltas: true,
        structured_stop_reasons: false,
        supports_assistant_prelude_before_tool_use: true,
        tool_call_args_encoding: ToolCallArgsEncoding::JsonString,
        tool_result_role_style: ToolResultRoleStyle::OpenaiToolRole,
    }
}

/// Resolve the thinking parity strategy given the protocol and request.
fn resolve_thinking_strategy(
    requested_thinking_disabled: bool,
    protocol: &str,
) -> ThinkingParityStrategy {
    if protocol == "mossen-compatible" {
        return ThinkingParityStrategy::Native;
    }

    if requested_thinking_disabled {
        return ThinkingParityStrategy::None;
    }

    if let Ok(env_strategy) = std::env::var("MOSSEN_CODE_CUSTOM_THINKING_PARITY_STRATEGY") {
        let trimmed = env_strategy.trim();
        if trimmed == "synthetic_two_pass" {
            return ThinkingParityStrategy::SyntheticTwoPass;
        }
        if trimmed == "none" {
            return ThinkingParityStrategy::None;
        }
    }

    ThinkingParityStrategy::SyntheticSinglePass
}

/// Get the official semantic capabilities for the current backend protocol.
pub fn get_official_semantic_capabilities(protocol: &str) -> OfficialSemanticCapabilities {
    if protocol == "openai-compatible" {
        openai_compatible_capabilities()
    } else {
        mossen_compatible_capabilities()
    }
}

/// Resolve the provider model policy for a given request.
pub fn resolve_provider_model_policy(
    protocol: &str,
    requested_thinking_disabled: bool,
) -> ProviderModelPolicy {
    ProviderModelPolicy {
        capabilities: if protocol == "openai-compatible" {
            openai_compatible_capabilities()
        } else {
            mossen_compatible_capabilities()
        },
        synthetic_tags: SyntheticTags::default(),
        thinking_strategy: resolve_thinking_strategy(requested_thinking_disabled, protocol),
    }
}
